### ISSUE TITLE
Support Filter Object

### ISSUE THREAD

**Issue by kanitw** [vega/vega-lite#1394]
Forking from #1376.   We are considering filter object syntax 

```
"filter": [
  {"field": "foo", "equal": "a", "negate": false},   // I want equal separate from in because then we can think about CompassQL's syntax for enumerating the drill down value.  
  {"field": "foo", "in": ["a", "b", "c"], "negate": false},
  {"field": "bar", "min": 5, "max": 10, "negate": false}
];
```

For backward compatability, we *might* want to support Vega Expression `string` or something like this:

```
// full expression version 
"filter": [
  {"test": "datum.foo == 'Kanit'"}
]
```

I don't have a strong opinion whether we should support the `"test": __expr__` version though.

**Comment by mprudhom** [comment]
Here's what we are using for our tool's pre-vega-lite data filter. It would be great if we could harmonize on syntax.

`fieldName >= 50`:
```json
{ "binary": "gte", "lhs": { "field": "fieldName" }, "rhs": { "value": 50 } }
```

`fieldName >= 50 && fieldName < 80`:
```json
{
  "binary": "and",
  "lhs": {
    "binary": "gte",
    "lhs": {"field": "fieldName"},
    "rhs": {"value": 50}
  },
  "rhs": {
    "binary": "lt",
    "lhs": {"field": "fieldName"},
    "rhs": {"value": 80}
  }
}
```

Basically, most operations are binary and take a left-hand side and a right-hand side. Compound predicates like `A || B || C` are represented as a binary "or" operation that nests a second binary "or" operation on the right-hand side:

```json
{
  "binary": "or",
  "lhs": {"field": "A"},
  "rhs": {"binary": "or","lhs": {"field": "B"},"rhs": {"field": "C"}}
}
```

Functions are represented with sub-expressions, so something like `weekday(A) != "Monday"` is represented like:

```json
{ "binary": "ne", "lhs": { "unary": "dow", "exp": { "field": "A" } }, "rhs": { "value": "Monday" } }
```

Similarly for arithmetic functions, where `A + 10 < B - 5` is expressed as:

```json
{
  "binary": "lt",
  "lhs": {"binary": "add","lhs": {"field": "A"},"rhs": {"value": 10}},
  "rhs": {"binary": "sub","lhs": {"field": "B"},"rhs": {"value": 5}}
}
```

Negation is just another function, so `!(A == 5)` is:

```json
{
  "unary": "not",
  "exp": {"binary": "eq","lhs": {"field": "A"},"rhs": {"value": 5}}
}
```

Keeping the language limited to mostly unary and binary operations and sub-expressions makes the language quite simple. I can elaborate on it in much more detail if there is interest. We have it compiling to SQL now, and it is pretty well tested. Our plan is to have it able to run against SQL for large remote data sources or directly to vega filter expressions for small in-memory data, but if we could instead use it directly in vega-lite that would be ideal.

FTR, we searched around for quite a while to see if there where any other projects that have worked out a general filter syntax JSON schema, and we were surprised that we haven't really found anything. mongoDB's JSON query is nice and compact, but using field names as keys would make it challenging to express the language in a JSON schema, which is a requirement for us.

**Comment by kanitw** [comment]
@mprudhom  Thanks for your input.  However, we do not want the filter object to be AST-like as it's too complicated for general use and make it much harder to read by people.  (Machine-wise, it doesn't matter.)

**Comment by mprudhom** [comment]
If that decision has been made, then OK. But I do think that you are going to wind up having to write something that is pretty close to an AST if you want complete expressive flexibility. For example with the following query:

```json
"filter": [
  {"field": "foo", "equal": "a", "negate": false}, 
  {"field": "foo", "in": ["a", "b", "c"], "negate": false},
  {"field": "bar", "min": 5, "max": 10, "negate": false}
]
```

Presumably the "filter" array is an AND expression. What if you want an OR expression, or an AND nested inside an OR?

```json
  {"field": "foo", "equal": "a", "negate": false}, 
```

What if you want to check whether the field is equal to another field instead of a constant value? Or if you want to compare the lowercase value of the 'foo' field to "a"?

```json
  {"field": "bar", "min": 5, "max": 10, "negate": false}
```

Is that inclusive or exclusive of the min/max values? What if you want the min exclusive and the max inclusive?

I don't mean to pick apart your specific example (which I assume was just off the top of your head), but I do want to illustrate that if one of the goals is to allow the composition of arbitrarily expressive filters, then you're probably going to wind up having something that looks like an AST.

**Comment by kanitw** [comment]
Thanks.  To clarify, the syntax above isn't perfectly set in stone but we don't want this syntax to be as complicated as you proposed.  (In fact, that looks very similar to our earlier draft that we think are too complicated.) 

> What if you want an OR expression, or an AND nested inside an OR?

To be fair, the `and` and `or` tree like this is reasonable.  (and/or do not always have to be binary)

```
{
   or: [
      {and: [ Predicate, ...] },
      Predicate
   ]
}
```
However, the default should be `and` because majority of people would use that.  Consider quick filter in tools like Tableau, users would add many filters that are and-ed together.  

> What if you want to check whether the field is equal to another field instead of a constant value?

That also rarely happen compared to general use case.  In any case, you can always write a custom expression or calculate a new field `aEqB` = `datum.a === datum.b` and simply do `{field: 'aEqB', equal: true}`.  It's not worth making general use case much more complicated to support this rare case.  

> Is that inclusive or exclusive of the min/max values? What if you want the min exclusive and the max inclusive?

Good point.  We can replace `min` with `>/>=` or `gt/gte`and it will be clearer.  

```
{"field": "bar", ">": 5, "<": 10}
```

> if you want complete expressive flexibility

Expressiveness isn't the only goal.  Another (more) important goal is also making common use case very easy.  In any case, people can always write the Vega expression for super complicated stuff.

**Comment by mprudhom** [comment]
> Expressiveness isn't the only goal. Another (more) important goal is also making common use case very easy. In any case, people can always write the Vega expression for super complicated stuff.

I would actually argue the reverse: for simple filters, it will always be easier for most users to just compose a vega expression like `{ "test": "bar >= 5 && bar < 10" }` then it will be for them to remember the keys for composing a query like `{"field": "bar", ">=": 5, "<": 10}`. I think it is when the filters start becoming more complex that editors will benefit from being able to provide rich structure for the query.

Anyhow, as I mentioned, I am coming at it from the tool production/consumption angle, so I am biased towards a more structured approach. But I'll refrain from further alternative proposals.

**Comment by kanitw** [comment]
Thank you for all comments.  I wanna make it clear that we welcome and appreciate your comments for this and other issues.  However, we disagree with *some* of the previous proposals (and agree with some parts as explained above, with variant of syntax the we prefer more).

The whole goal of filter object is provide structure to support higher-level tools like you mentioned.  
However, we cannot think of Vega-Lite specs as only handwritten or only machine-generated.  

Vega-Lite as a specification would be a part of an eco-system.   Machine generated specs might be exported and shared, thus should be human-readable as much as possible.   

Therefore,  we would favor structure like this

```
{field: "fieldName", "moreThan/lessThan": 50}   // <, > actually don't work due to TS restriction
```

rather than

```
{
  "binary": "gt/lt",
  "lhs": {"field": "fieldName"},
  "rhs": {"value": 50}
}
```

because the former is human-friendlier if you share it with other people (with trade-offs that some rare case aren't expressible). 

For the same reason, we don't want the filter object to include arbitrary calculation as you can always use calculate / Vega expression to achieve the same result.  

I do agree that if you write by hand, the expression syntax is generally more convenient for Javascript users than the object form we are proposing (except when you start wanna do things with timeUnit, etc.). 
The filter object here aims to provide **human-readable** structure for higher-level tool for more common used case.  Plus, people who use Vega-Lite via [wrappers in other language](https://github.com/ellisonbg/altair), our JSON syntax also provides structure for those wrappers to implement human-friendly API as well.  

For less common use case, you can use the Vega expression and use the Vega-Expression version and [parse](https://github.com/vega/vega-expression#api-usage) the expression to make an AST in the form of [ESTree](https://github.com/estree/estree).  (This might be interesting to you.)

I hope I don't overlook anything, but feel free to reply if there are some unclear parts.

**Comment by kanitw** [comment]
Fixed in #1461
